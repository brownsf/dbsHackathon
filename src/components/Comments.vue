<template>
  <div class="commentContainer">
    <v-card tile>
      <v-toolbar
        card
        dark
        color="primary">
        <v-btn
          icon
          dark
          @click="$emit('close')">
          <v-icon>close</v-icon>
        </v-btn>
        <v-toolbar-title>Comments for {{ singleTopic.name }}</v-toolbar-title>
        <v-spacer/>
        <v-toolbar-items>
          <v-btn
            dark
            flat
            @click.native="dialog = true">add</v-btn>
        </v-toolbar-items>
      </v-toolbar>
      <v-container
        grid-list-md
        text-xs-center>
        <v-layout
          row
          wrap>
          <v-flex xs12>
            <v-list subheader>
              <v-subheader>Recent comments:</v-subheader>
              <v-progress-linear
                v-if="loading"
                class="loader" />
              <v-alert
                v-if="error"
                class="errorAlert">Error loading topic details</v-alert>
              <template v-for="(item, index) in singleTopic.comments">
                <v-list-tile

                  :key="item.id">
                  <v-list-tile-content>
                    <v-list-tile-title v-html="item.message"/>
                  </v-list-tile-content>
                </v-list-tile>
                <v-divider :key="`divider-${index}`"/>
              </template>
            </v-list>
          </v-flex>
        </v-layout>
      </v-container>
      <v-dialog
        v-model="dialog"
        max-width="500px">
        <v-card>
          <v-card-title>
            Add Comment
          </v-card-title>
          <v-form ref="form">
            <v-card-text>
              <v-layout row>
                <v-flex
                  xs12
                  sm8
                  offset-sm2>
                  <v-text-field
                    v-model="comment"
                    :rules="[v=>!!v||'Comment is required']"
                    label="Comment"
                    validate-on-blur
                    multi-line
                  />
                </v-flex>
              </v-layout>
            </v-card-text>
            <v-card-actions>
              <v-btn
                color="primary"
                class="saveComment"
                flat
                @click="saveComment">Add</v-btn>

              <v-btn
                color="secondary"
                flat
                @click.stop="dialog=false">Cancel</v-btn>
            </v-card-actions>
          </v-form>
        </v-card>
      </v-dialog>
    </v-card>
  </div>
</template>
<style>
.commentContainer {
  width: 100%;
  background: rgba(255, 255, 255, 0.7);
}
</style>

<script>
import { mapState, mapGetters } from 'vuex';

export default {
  props: {
    selected: {
      type: Number,
      default: () => 0,
    },
  },
  data: () => ({
    dialog: false,
    comment: '',
  }),
  computed: {
    ...mapState({
      loading: state => state.singleTopicLoad,
      error: state => state.singleTopicError,
    }),
    ...mapGetters(['singleTopic']),
  },
  watch: {
    selected: {
      handler(newVal) {
        // watch it
        this.getTopic(newVal);
      },
    },
  },
  methods: {
    getTopic(id) {
      if (id) {
        this.$store.dispatch('getTopic', id);
      }
    },
    saveComment() {
      if (this.$refs.form.validate()) {
        this.$store.dispatch('addComment', {
          topic_id: this.singleTopic.id,
          message: this.comment,
        }).then(() => {
          this.dialog = false;
        });
      }
    },
  },
};
</script>
