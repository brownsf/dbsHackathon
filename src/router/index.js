import Vue from 'vue';
import Router from 'vue-router';
import Register from '@/components/Register';
import Login from '@/components/Login';
import Topics from '@/components/Topics';
import Main from '@/components/Main';

Vue.use(Router);

export default new Router({
  routes: [
    {
      path: '/',
      name: 'Main',
      component: Main,
    },
    {
      path: '/register',
      name: 'Register',
      component: Register,
    },
    {
      path: '/login',
      name: 'Login',
      component: Login,
    },
    {
      path: '/topics',
      name: 'Topics',
      component: Topics,
    },
  ],
});
