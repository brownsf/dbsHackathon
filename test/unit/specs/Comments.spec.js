import { mount, createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import Vue from 'vue';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import Comments from '@/components/Comments';
import router from '@/router';

Vue.use(Vuex);
Vue.use(Vuetify);
Vue.use(VueRouter);
describe('Comments', () => {
  const state = { data: { topic: { name: 'here' }, singleTopicLoad: false } };
  const actions = {
    register: jest.fn(),
    getAllTopics: jest.fn(),
    login: jest.fn(),
    getTopic: jest.fn(),
    addComment: jest.fn(),
  };
  const getters = {
    isAuth: () => true,
    singleTopic: () => ({
      comments: [
        {
          message: 'one',
        },
        { message: 'two' },
      ],
    }),
  };

  const store = new Vuex.Store({
    state,
    actions,
    getters,
  });
  const wrapper = mount(Comments, { store, router });
  it('should render', () => {
    // expect(wrp.html()).toMatchSnapshot();

    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should get all topic info', () => {
    wrapper.setProps({ selected: 1 });
    expect(actions.getTopic.mock.calls).toHaveLength(1);
    wrapper.setProps({ selected: 0 });
    expect(actions.getTopic.mock.calls).toHaveLength(1);
  });

  it('should save comment', () => {
    wrapper.vm.$refs.form.validate = () => true;
    wrapper.find('.saveComment').trigger('click');
    expect(actions.addComment.mock.calls).toHaveLength(1);

    wrapper.vm.$refs.form.validate = () => false;
    wrapper.find('.saveComment').trigger('click');
    expect(actions.addComment.mock.calls).toHaveLength(1);
  });

  it('should render state corrently', () => {
    expect(wrapper.findAll('.loader')).toHaveLength(0);
    expect(wrapper.findAll('.errorAlert')).toHaveLength(0);
  });
});
