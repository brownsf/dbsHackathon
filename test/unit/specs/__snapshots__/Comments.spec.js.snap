// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments should render 1`] = `
<div class="commentContainer">
  <div class="card card--tile" style="height: auto;">
    <nav class="toolbar toolbar--card theme--dark teal" style="margin-top: 0px; padding-right: 0px; padding-left: 0px;">
      <div class="toolbar__content" style="height: 56px;">
        <button type="button" class="btn btn--icon theme--dark">
          <div class="btn__content">
            <i aria-hidden="true" class="icon material-icons">close</i>
          </div>
        </button>
        <div class="toolbar__title">Comments for </div>
        <div class="spacer"></div>
        <div class="toolbar__items">
          <button type="button" class="btn btn--flat theme--dark">
            <div class="btn__content">add</div>
          </button>
        </div>
      </div>
    </nav>
    <div class="container grid-list-md text-xs-center">
      <div class="layout row wrap">
        <div class="flex xs12">
          <div class="list list--subheader">
            <div class="subheader">Recent comments:</div>
            <!---->
            <!---->
            <div>
              <div class="list__tile">
                <div class="list__tile__content">
                  <div class="list__tile__title">one</div>
                </div>
              </div>
            </div>
            <hr class="divider">
            <div>
              <div class="list__tile">
                <div class="list__tile__content">
                  <div class="list__tile__title">two</div>
                </div>
              </div>
            </div>
            <hr class="divider">
          </div>
        </div>
      </div>
    </div>
    <div class="dialog__container" style="display: block;">
      <div class="dialog__content" tabindex="-1" style="z-index: 0;">
        <div class="dialog" style="max-width: 500px; display: none;">
          <div class="card" style="height: auto;">
            <div class="card__title">
              Add Comment
            </div>
            <form novalidate="novalidate">
              <div class="card__text">
                <div class="layout row">
                  <div class="flex xs12 sm8 offset-sm2">
                    <div class="input-group input-group--text-field input-group--multi-line primary--text">
                      <label>Comment</label>
                      <div class="input-group__input">
                        <textarea tabindex="0" aria-label="Comment" rows="5"></textarea>
                      </div>
                      <div class="input-group__details">
                        <!---->
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="card__actions">
                <button type="button" class="saveComment btn btn--flat primary--text">
                  <div class="btn__content">Add</div>
                </button>
                <button type="button" class="btn btn--flat secondary--text">
                  <div class="btn__content">Cancel</div>
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
